File: ADP.Surveys/ADP.Surveys.Shared/Evaluation/ExpressionSandbox/Lexer.cs
using System.Globalization;
using System.Text;

namespace ShiftSoftware.ADP.Surveys.Shared.Evaluation.ExpressionSandbox;

public enum TokenKind
{
    Number, String, True, False, Null,
    Identifier,
    Eq, StrictEq, NotEq, StrictNotEq,
    Lt, Gt, LtEq, GtEq,
    And, Or, Not,
    LParen, RParen, LBracket, RBracket,
    Comma, Dot,
    EndOfInput,
}

public readonly record struct Token(TokenKind Kind, string Text, object? Literal, int Position);

public class ExpressionSyntaxException(string message, int position)
    : Exception($"Expression syntax error at column {position}: {message}")
{
    public int Position { get; } = position;
}

/// <summary>
/// Hand-rolled lexer for the survey expression mini-language. Kept intentionally
/// terse — this is the shape the TypeScript SDK will mirror.
/// </summary>
internal sealed class Lexer(string source)
{
    private readonly string source = source;
    private int pos;

    public List<Token> Tokenize()
    {
        var tokens = new List<Token>();
        while (true)
        {
            SkipWhitespace();
            if (pos >= source.Length)
            {
                tokens.Add(new Token(TokenKind.EndOfInput, "", null, pos));
                return tokens;
            }
            tokens.Add(NextToken());
        }
    }

    private void SkipWhitespace()
    {
        while (pos < source.Length && char.IsWhiteSpace(source[pos])) pos++;
    }

    private Token NextToken()
    {
        var start = pos;
        var c = source[pos];

        if (char.IsDigit(c)) return ReadNumber(start);
        if (c == '\'' || c == '"') return ReadString(start, c);
        if (c == '_' || char.IsLetter(c)) return ReadIdentifier(start);

        switch (c)
        {
            case '(': pos++; return new(TokenKind.LParen, "(", null, start);
            case ')': pos++; return new(TokenKind.RParen, ")", null, start);
            case '[': pos++; return new(TokenKind.LBracket, "[", null, start);
            case ']': pos++; return new(TokenKind.RBracket, "]", null, start);
            case ',': pos++; return new(TokenKind.Comma, ",", null, start);
            case '.': pos++; return new(TokenKind.Dot, ".", null, start);
            case '=':
                if (Match("===")) return new(TokenKind.StrictEq, "===", null, start);
                if (Match("==")) return new(TokenKind.Eq, "==", null, start);
                throw new ExpressionSyntaxException("bare '=' is not a valid operator (use '==' or '===').", start);
            case '!':
                if (Match("!==")) return new(TokenKind.StrictNotEq, "!==", null, start);
                if (Match("!=")) return new(TokenKind.NotEq, "!=", null, start);
                pos++; return new(TokenKind.Not, "!", null, start);
            case '<':
                if (Match("<=")) return new(TokenKind.LtEq, "<=", null, start);
                pos++; return new(TokenKind.Lt, "<", null, start);
            case '>':
                if (Match(">=")) return new(TokenKind.GtEq, ">=", null, start);
                pos++; return new(TokenKind.Gt, ">", null, start);
            case '&':
                if (Match("&&")) return new(TokenKind.And, "&&", null, start);
                throw new ExpressionSyntaxException("expected '&&'.", start);
            case '|':
                if (Match("||")) return new(TokenKind.Or, "||", null, start);
                throw new ExpressionSyntaxException("expected '||'.", start);
        }

        throw new ExpressionSyntaxException($"unexpected character '{c}'.", start);
    }

    private bool Match(string s)
    {
        if (pos + s.Length > source.Length) return false;
        for (var i = 0; i < s.Length; i++)
            if (source[pos + i] != s[i]) return false;
        pos += s.Length;
        return true;
    }

    private Token ReadNumber(int start)
    {
        while (pos < source.Length && char.IsDigit(source[pos])) pos++;
        if (pos < source.Length && source[pos] == '.')
        {
            pos++;
            while (pos < source.Length && char.IsDigit(source[pos])) pos++;
        }
        var text = source.Substring(start, pos - start);
        var value = decimal.Parse(text, CultureInfo.InvariantCulture);
        return new(TokenKind.Number, text, value, start);
    }

    private Token ReadString(int start, char quote)
    {
        pos++; // opening quote
        var sb = new StringBuilder();
        while (pos < source.Length && source[pos] != quote)
        {
            var c = source[pos];
            if (c == '\\' && pos + 1 < source.Length)
            {
                var next = source[pos + 1];
                sb.Append(next switch
                {
                    'n' => '\n',
                    't' => '\t',
                    'r' => '\r',
                    '\\' => '\\',
                    '\'' => '\'',
                    '"' => '"',
                    _ => throw new ExpressionSyntaxException($"unknown escape '\\{next}'.", pos),
                });
                pos += 2;
            }
            else
            {
                sb.Append(c);
                pos++;
            }
        }
        if (pos >= source.Length)
            throw new ExpressionSyntaxException("unterminated string literal.", start);
        pos++; // closing quote
        var text = sb.ToString();
        return new(TokenKind.String, text, text, start);
    }

    private Token ReadIdentifier(int start)
    {
        while (pos < source.Length && (source[pos] == '_' || char.IsLetterOrDigit(source[pos]) || source[pos] == '-'))
            pos++;
        var text = source.Substring(start, pos - start);
        return text switch
        {
            "true" => new(TokenKind.True, text, true, start),
            "false" => new(TokenKind.False, text, false, start),
            "null" => new(TokenKind.Null, text, null, start),
            _ => new(TokenKind.Identifier, text, null, start),
        };
    }
}

File: ADP.Surveys/renderer/packages/survey-sdk/src/expression-sandbox/lexer.ts
/**
 * Hand-rolled lexer for the survey expression mini-language. Mirror of
 * `ADP.Surveys.Shared/Evaluation/ExpressionSandbox/Lexer.cs`.
 */

export type TokenKind =
  | 'Number'
  | 'String'
  | 'True'
  | 'False'
  | 'Null'
  | 'Identifier'
  | 'Eq'
  | 'StrictEq'
  | 'NotEq'
  | 'StrictNotEq'
  | 'Lt'
  | 'Gt'
  | 'LtEq'
  | 'GtEq'
  | 'And'
  | 'Or'
  | 'Not'
  | 'LParen'
  | 'RParen'
  | 'LBracket'
  | 'RBracket'
  | 'Comma'
  | 'Dot'
  | 'EndOfInput';

export interface Token {
  kind: TokenKind;
  text: string;
  literal: number | string | boolean | null;
  position: number;
}

export class ExpressionSyntaxError extends Error {
  readonly position: number;
  constructor(message: string, position: number) {
    super(`Expression syntax error at column ${position}: ${message}`);
    this.position = position;
    this.name = 'ExpressionSyntaxError';
  }
}

function isDigit(c: string): boolean {
  return c >= '0' && c <= '9';
}

function isLetter(c: string): boolean {
  return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z');
}

function isLetterOrDigit(c: string): boolean {
  return isLetter(c) || isDigit(c);
}

function isWhitespace(c: string): boolean {
  return c === ' ' || c === '\t' || c === '\n' || c === '\r' || c === '\f' || c === '\v';
}

export function tokenize(source: string): Token[] {
  const tokens: Token[] = [];
  let pos = 0;

  const atEnd = (): boolean => pos >= source.length;
  const peekAt = (offset = 0): string => source.charAt(pos + offset);

  const match = (s: string): boolean => {
    if (pos + s.length > source.length) return false;
    for (let i = 0; i < s.length; i++) {
      if (source.charAt(pos + i) !== s.charAt(i)) return false;
    }
    pos += s.length;
    return true;
  };

  const skipWhitespace = (): void => {
    while (!atEnd() && isWhitespace(peekAt())) pos++;
  };

  const readNumber = (start: number): Token => {
    while (!atEnd() && isDigit(peekAt())) pos++;
    if (!atEnd() && peekAt() === '.') {
      pos++;
      while (!atEnd() && isDigit(peekAt())) pos++;
    }
    const text = source.substring(start, pos);
    const value = parseFloat(text);
    return { kind: 'Number', text, literal: value, position: start };
  };

  const readString = (start: number, quote: string): Token => {
    pos++; // opening quote
    let sb = '';
    while (!atEnd() && peekAt() !== quote) {
      const c = peekAt();
      if (c === '\\' && pos + 1 < source.length) {
        const next = peekAt(1);
        const unescaped: Record<string, string> = {
          n: '\n',
          t: '\t',
          r: '\r',
          '\\': '\\',
          "'": "'",
          '"': '"',
        };
        const replacement = unescaped[next];
        if (replacement === undefined)
          throw new ExpressionSyntaxError(`unknown escape '\\${next}'.`, pos);
        sb += replacement;
        pos += 2;
      } else {
        sb += c;
        pos++;
      }
    }
    if (atEnd()) throw new ExpressionSyntaxError('unterminated string literal.', start);
    pos++; // closing quote
    return { kind: 'String', text: sb, literal: sb, position: start };
  };

  const readIdentifier = (start: number): Token => {
    while (!atEnd()) {
      const c = peekAt();
      if (c === '_' || c === '-' || isLetterOrDigit(c)) pos++;
      else break;
    }
    const text = source.substring(start, pos);
    if (text === 'true') return { kind: 'True', text, literal: true, position: start };
    if (text === 'false') return { kind: 'False', text, literal: false, position: start };
    if (text === 'null') return { kind: 'Null', text, literal: null, position: start };
    return { kind: 'Identifier', text, literal: null, position: start };
  };

  const nextToken = (): Token => {
    const start = pos;
    const c = peekAt();

    if (isDigit(c)) return readNumber(start);
    if (c === "'" || c === '"') return readString(start, c);
    if (c === '_' || isLetter(c)) return readIdentifier(start);

    switch (c) {
      case '(':
        pos++;
        return { kind: 'LParen', text: '(', literal: null, position: start };
      case ')':
        pos++;
        return { kind: 'RParen', text: ')', literal: null, position: start };
      case '[':
        pos++;
        return { kind: 'LBracket', text: '[', literal: null, position: start };
      case ']':
        pos++;
        return { kind: 'RBracket', text: ']', literal: null, position: start };
      case ',':
        pos++;
        return { kind: 'Comma', text: ',', literal: null, position: start };
      case '.':
        pos++;
        return { kind: 'Dot', text: '.', literal: null, position: start };
      case '=':
        if (match('===')) return { kind: 'StrictEq', text: '===', literal: null, position: start };
        if (match('==')) return { kind: 'Eq', text: '==', literal: null, position: start };
        throw new ExpressionSyntaxError(
          "bare '=' is not a valid operator (use '==' or '===').",
          start,
        );
      case '!':
        if (match('!==')) return { kind: 'StrictNotEq', text: '!==', literal: null, position: start };
        if (match('!=')) return { kind: 'NotEq', text: '!=', literal: null, position: start };
        pos++;
        return { kind: 'Not', text: '!', literal: null, position: start };
      case '<':
        if (match('<=')) return { kind: 'LtEq', text: '<=', literal: null, position: start };
        pos++;
        return { kind: 'Lt', text: '<', literal: null, position: start };
      case '>':
        if (match('>=')) return { kind: 'GtEq', text: '>=', literal: null, position: start };
        pos++;
        return { kind: 'Gt', text: '>', literal: null, position: start };
      case '&':
        if (match('&&')) return { kind: 'And', text: '&&', literal: null, position: start };
        throw new ExpressionSyntaxError("expected '&&'.", start);
      case '|':
        if (match('||')) return { kind: 'Or', text: '||', literal: null, position: start };
        throw new ExpressionSyntaxError("expected '||'.", start);
    }

    throw new ExpressionSyntaxError(`unexpected character '${c}'.`, start);
  };

  while (true) {
    skipWhitespace();
    if (atEnd()) {
      tokens.push({ kind: 'EndOfInput', text: '', literal: null, position: pos });
      return tokens;
    }
    tokens.push(nextToken());
  }
}
